src/Microsoft.OData.Edm/Csdl/Parsing/SchemaJsonParser.cs

```diff
@@ -0,0 +1,1564 @@
+        {
+            navigationSource = null;
+            if (element.ValueKind != JsonValueKind.Object)
+            {
+                return false;
+            }
+
+            IList<CsdlNavigationPropertyBinding> navigationPropertyBindings = new List<CsdlNavigationPropertyBinding>();
+            IList<CsdlAnnotation> annotations = new List<CsdlAnnotation>();
+            string type = null;
+            bool? isCollection = null;
+            bool? includeInServiceDocument = null;
+            bool? nullable = null;
+            element.ParseAsObject(context, (propertyName, propertyValue) =>
+            {
+                switch (propertyName)
+                {
+                    case "$Type":
+                        // The value of $Type is the qualified name of an entity type.
+                        type = propertyValue.ParseAsString(context);
+                        break;
+
+                    case "$Collection":
+                        // The value of $Collection is the Booelan value true.
+                        isCollection = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    case "$IncludeInServiceDocument":
+                        // The value of $IncludeInServiceDocument is one of the Boolean literals true or false. Absence of the member means true.
+                        includeInServiceDocument = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    case "$Nullable":
+                        // The value of $Nullable is one of the Boolean literals true or false. Absence of the member means false.
+                        // In OData 4.0 responses this member MUST NOT be specified.
+                        nullable = propertyValue.ParseAsBoolean(context);
+                        break;
+
+                    case "$NavigationPropertyBinding":
+                        // The value of $NavigationPropertyBinding is an object.
+                        navigationPropertyBindings = ParseCsdlNavigationPropertyBinding(propertyValue, context);
+                        break;
+
+                    default:
+                        // It may contain annotation
+                        ParseCsdlAnnotation(propertyName, propertyValue, context, annotations);
+                        break;
+                }
+            });
+
+            if (type == null)
+            {
+                return false;
+            }
+
+            CsdlLocation location = context.Location();
+
+            if (isCollection != null)
+            {
+                // entitySet
+                if (!isCollection.Value)
+                {
+                    // If presented, $IsCollection should be true
+                    context.ReportError(EdmErrorCode.JsonInvalidCollectionValue, "$IsCollection should be true");
+                }

```

Comment:
why? should be able to have a singleton specifying $IsCollection=false

Reply:
OData spec says
For collection-valued properties the value of $Type is the qualified name of the property’s item type, and the member $Collection MUST be present with the literal value true.

So, in my understanding, $Collection only presents if the element is a collection, and its value should be "true".
And that's the only way to distinguish "EntitySet" and "Singleton".

Because, in "EntitySet", OData spec has:

```
$Collection

The value of $Collection is the Booelan value true.
```

But, Singleton doesn't have "$Collection".



---
In reply to: [443067475](https://github.com/OData/odata.net/pull/1775#discussion_r443067475) [](ancestors = 443067475)